A52-xds-custom-lb-policies.md

```diff
@@ -0,0 +1,315 @@
+
+## Proposal
+
+Load balancing configuration of a gRPC client is based on the fields in the [`Cluster`](https://github.com/envoyproxy/envoy/blob/305a200ce2002de9984e7fb7bb98d92b8e9fbcca/api/envoy/config/cluster/v3/cluster.proto#L48) xDS resource delivered to the client by the [`AggregatedDiscoveryService`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/discovery/v3/ads.proto.html). Currently the combination of the [`lb_policy`](https://github.com/envoyproxy/envoy/blob/55a506f4cc053c9841c44644238aced88dea191d/api/envoy/config/cluster/v3/cluster.proto#L779) enum value and a policy specific configuration field in the [`lb_config`](https://github.com/envoyproxy/envoy/blob/305a200ce2002de9984e7fb7bb98d92b8e9fbcca/api/envoy/config/cluster/v3/cluster.proto#L991) oneof are used to determine the policy configuration. What will be added is the additional support for the newer [`load_balancing_policy`](https://github.com/envoyproxy/envoy/blob/305a200ce2002de9984e7fb7bb98d92b8e9fbcca/api/envoy/config/cluster/v3/cluster.proto#L1064) field.
+
+If the `load_balancing_policy` field is provided, it will be used instead of the `lb_policy` and `lb_config` fields. If `load_balancing_policy` is not provided, full backward compatibility with the old fields will be maintained.
+
+### XdsClient
+
+When the xDS client receives a CDS update it notifies its listeners/watchers of the configuration changes, including the load balancing policy to be used. Currently the xDS client only considers the `lb_policy` field in the CDS update when constructing the update information it sends out. This logic will change to consider the new `load_balancing_policy` field that will be used as the source of LB policy configuration if it is populated.
+
+Regardless of the field that load balancing policy configuration is picked up from, internally `XdsClient` will represent the information in a uniform manner. The CDS update struct will be updated to carry LB policy configuration in JSON form. We move from an enum for the policy selection and a set of `oneof` configuration fields to having a [LoadBalancingConfig](https://github.com/grpc/grpc-proto/blob/ab96cf12ec7ce135e03d6ea91d96213fa4cb02af/grpc/service_config/service_config.proto#L560) JSON object.
+
+The load balancer config portion of the CDS update struct will change to take the following form:
+
+```c
+{
+  ...
+
+  "lb_policy_config": <array of LoadBalancingConfig JSON objects>
+  
+  ...
+}
+```
+
+The XdsClient will transfer the responsibility of converting the xDS configuration to the internal JSON format to a new *xDS LB policy registry*. This new registry will return an error if the xDS load balancing configuration cannot be converted, but it will not validate a converted configuration. It will be the responsibility of the XdsClient to validate that the converted configuration. It will do this by having the gRPC LB policy registry parse the configuration.
```

Comment:
```suggestion
The XdsClient will transfer the responsibility of converting the xDS configuration to the internal JSON format to a new *xDS LB policy registry*. This new registry will return an error if the xDS load balancing configuration cannot be converted, but it will not validate a converted configuration. It will be the responsibility of the XdsClient to validate the converted configuration. It will do this by having the gRPC LB policy registry parse the configuration.
```